common/test/test_throttle.cpp

```diff
@@ -161,11 +165,266 @@ TEST(Throttle, try_consume) {
+));
+
+////////////////////////////////////////
+
+struct PriorityTestSuite {
+    enum Type {
+        Simulate,
+        RealSocket,
+    };
+    struct IOConfig {
+        uint64_t bw;    // bandwidth per second
+        uint64_t bs;    // block size per IO
+        photon::throttle::Priority prio;
+    };
+
+    Type type;
+    uint64_t limit_bw;
+    IOConfig io1;
+    IOConfig io2;
+    double bw1_ratio_min, bw1_ratio_max;
+    double bw2_ratio_min, bw2_ratio_max;
+};
+
+class ThrottlePriorityTest : public testing::TestWithParam<PriorityTestSuite> {
+};
+
+INSTANTIATE_TEST_CASE_P(Throttle, ThrottlePriorityTest, testing::Values(
+        PriorityTestSuite{
+                PriorityTestSuite::Simulate,
+                100'000'000,
+                {50'000'000, 100'000, photon::throttle::Priority::High},
+                {50'000'000, 100'000, photon::throttle::Priority::High},
+                0.4, 0.6,
+                0.4, 0.6,
+        },
+        PriorityTestSuite{
+                PriorityTestSuite::Simulate,
+                100'000'000,
+                {50'000'000, 1'000'000, photon::throttle::Priority::High},
+                {150'000'000, 2'000'000, photon::throttle::Priority::High},
+                0.4, 0.6,
+                0.4, 0.6,
+        },
+        PriorityTestSuite{
+                PriorityTestSuite::Simulate,
+                100'000'000,
+                {100'000'000, 500'000, photon::throttle::Priority::High},
+                {100'000'000, 500'000, photon::throttle::Priority::Low},
+                0.9, 1.0,
+                0.0, 0.1,
+        },
+        PriorityTestSuite{
+                PriorityTestSuite::Simulate,
+                100'000'000,
+                {30'000'000, 1'000'000, photon::throttle::Priority::High},
+                {70'000'000, 1'000'000, photon::throttle::Priority::Low},
+                0.25, 0.35,
+                0.3, 0.7,
+        },
+        PriorityTestSuite{
+                PriorityTestSuite::Simulate,
+                100'000'000,
+                {50'000'000, 5'000'000, photon::throttle::Priority::High},
+                {200'000'000, 10'000'000, photon::throttle::Priority::Low},
+                0.4, 0.5,
+                0.5, 0.7,
+        },
+        PriorityTestSuite{
+                // For now there is no way to balance throttle throughput of the same priority.
+                PriorityTestSuite::RealSocket,
+                1'000'000'000,
+                {1'000'000'000, 1048576, photon::throttle::Priority::High},
+                {1'000'000'000, 1048576, photon::throttle::Priority::High},
+                0.0, 1.0,
+                0.0, 1.0,
+        },
+        PriorityTestSuite{
+                PriorityTestSuite::RealSocket,
+                1'000'000'000,
+                {800'000'000, 32768, photon::throttle::Priority::High},
+                {800'000'000, 32768, photon::throttle::Priority::Low},
+                0.8, 1.1,
```

Comment:
Why here high priority throughput should take 80% to 110% of whole limit? Since `io1` with 800M throttle while `src` with 1000M throttle, the ratio should be 80% at most, that makes expected ratio range hardly fulfilled.